rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+
+Pros:
+
+- Avoids introducing a new RPC and ensures that all data within the Heartbeat
+  comes from the same instance in time.
+- Allows self-reported information to be used as part of renewal decision.
+  This is not a strong defence as it is self-reported and cannot be trusted.
+- Avoids a state where the BotInstance is incomplete immediately after joining
+  and before it has called SubmitHeartbeat.
+
+Cons:
+
+- Adds Bot specific behaviour to RPCs that are also used for Node joining.
+- Heartbeats are limited to the interval of renewal.
+
+#### API
+
+Additional RPCs will be added to the BotInstance service to allow these to
+be listed and deleted:
+
+```protobuf
+syntax = "proto3";
+
+package teleport.machineid.v1;
+
+service BotInstanceService {
+  // GetBotInstance returns the specified BotInstance resource.
+  rpc GetBotInstance(GetBotInstanceRequest) returns (BotInstance);
+  // ListBotInstances returns a page of BotInstance resources.
+  rpc ListBotInstances(ListBotInstancesRequest) returns (ListBotInstancesResponse);
+  // DeleteBotInstance hard deletes the specified BotInstance resource.
+  rpc DeleteBotInstance(DeleteBotInstanceRequest) returns (google.protobuf.Empty);
+}
+
+// Request for GetBotInstance.
+message GetBotInstanceRequest {
+  // The name of the BotInstance to retrieve.
+  string name = 1;
+}
+
+// Request for ListBotInstances.
+//
+// Follows the pagination semantics of
+// https://cloud.google.com/apis/design/standard_methods#list
+message ListBotInstancesRequest {
+  // The name of the Bot to list BotInstances for. If empty, all BotInstances
+  // will be listed.
+  string filter_bot_name = 1;
+  // The maximum number of items to return.
+  // The server may impose a different page size at its discretion.
+  int32 page_size = 2;
+  // The page_token value returned from a previous ListBotInstances request, if
+  // any.
+  string page_token = 3;
+}
+
+// Response for ListBotInstances.
+message ListBotInstancesResponse {
+  // BotInstance that matched the search.
+  repeated BotInstance bot_instances = 1;
+  // Token to retrieve the next page of results, or empty if there are no
+  // more results exist.
+  string next_page_token = 2;
+}
+
+// Request for DeleteBotInstance.
+message DeleteBotInstanceRequest {
+  // The name of the BotInstance to delete.
+  string name = 1;
+}
+```
+
+### Changes to the `token` Join Method
+
+As we now have a way to track the generation for a specific Bot instance, we
+can allow multiple Bot instances to be associated with a single Bot. This
+also means that the token no longer needs to be consumed on a join.
+
+Eventually, we may wish to add a way to specify a number of joins which can
+occur with a token. This provides a way to easily control the lifetime of a 
+token when deploying to a fleet of a pre-known size.
+
+The renewal logic will need to be adjusted to read and update the generation
+counter from the BotInstance rather than the Bot user.
```

Comment:
> I'm a bit conflicted about launching multi-use tokens without also adding options to limit joins.

I agree some sort of limiting factor on a number of joins is likely going to be necessary out of the box.

> We should include a join count limit from the start, and arguably default it to 1 so users are making an explicit decision to allow multiple bots to join. This default would also avoid a potentially surprising change in behavior for existing users.

I like this idea - especially using a default of 1 to avoid a big change in behaviour.

---

One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

We'd still want them to rotate though... one thing I was thinking about to make this easier would be to produce a script or command which can be executed on a cron job that generates a new token for a named bot and then calls a user-supplied script which should connect to their CI platform/secret store and updates the token. This may be a nice answer for those who are backed into a corner by a rudimentary CI platform.

This fits well into the "bot refresh" idea, some sort of easyish CLI command where you specify an existing bot and can just generate a new token.

Reply:
> One use-case I think we can't neglect is CI platforms we do not support and can not support due to their limitations. In this case, users may well still want a token that is "quite" long-lived and supports a large number of joins.

I agree. I almost wonder if there's a middle ground here where we might be able to force users to make a security decision:
* The `token` join method sets a sane cap on the TTL of a join token to proactively encourage some variety of automation, say 7 days or something. I don't think we need to cap the number of uses as long as there's a time bound.
  * It might be useful to write up a cron job example, maybe a guide on using a GHA or k8s cronjob to rotate a Jenkins joining secret or something.
* We then (eventually) introduce a new `token-insecure` that otherwise behaves identically but has no restrictions at all on TTL or number of joins (basically this [other thread](https://github.com/gravitational/teleport/pull/36510/files#r1602429025) down the page). This method might also opt out of the generation counter.

I think we could argue enforcing a maximum token TTL is an acceptable breaking change given how narrowly useful long TTL'd tokens are today, and then introduce `token-insecure` at some point in the future? I can't imagine any users today are making practical use of extremely long TTL bot tokens that would have their use cases broken.

Reply:
I'm a little cautious about setting a cap on the TTL, if we were totally greenfield I'd agree but:
- Today, there's no cap. I agree with you and doubt anyone is relying on this behaviour - so perhaps we can discount this.
- We'd need to have logic that ignores the cap if it's for Agent joining
- We'd need to have logic that ignores the cap if the join method is not `token`.

I'd perhaps be less cautious about the TTL cap if we were approaching this at a later date when we add `token-insecure`, imho, the whole token type could do with a V3 and that might be a good opportunity to do so. As an aside, I'd love to make it so token name is not the actual secret value, e.g a `token` join method token has a UUID-esque name or a user provided name, and then a field which specifies the actual secret value. This would make audit logging etc much easier because we wouldn't need to "hide" the name.

I'm not overly decided on this and I think I need to ponder on it a little longer to figure out my full thoughts. It's almost out of scope for this RFD - so perhaps we don't need to make a decision on this immediately and we can proceed with the BotInstance work since I don't think this has any impact.

Reply:
I think my hopefully more in scope recommendation for this RFD is: let's enforce a soft client-side limit in `tctl` and/or warn that it's a bad idea and we may restrict token length in the future. I figure that if a user wants to create an extremely long-lived token, they can write up the YAML themselves and `tctl create -f` it.

I think warnings or soft errors today help keep the door open for us to lock this down more in the future while making sure use cases that need very long lived tokens are at least possible.